### PR TITLE
New decompressor API

### PR DESCRIPTION
**Pull request by yawkat**
Motivation:

The current decompression decoders have a few issues:

* They're bound to ChannelHandler, so for HTTP, they need to use an EmbeddedChannel with all the associated overhead.
* They do not support backpressure, so the zip bomb fix could be better.
* Each one has about 5 constructor overloads

Modification:

* Introduce a new `Decompressor` API not based on `ChannelHandler`. It is similar in design to the brotli API, with separate operations for adding input and removing output.
* Implement decompressors for all supported formats. The code is copied from the decoders, which remain untouched. No effort is made to reduce code duplication. This is why the PR is enormous.
* Use builders as a unified way of constructing all the new decompressors, instead of constructor overloads.
* Implement a `BackpressureDecompressionHandler` that uses the new `Decompressor` API.

Result:

Lower overhead, better backpressure handling.

Work yet to be done:

* Deprecate old decoders, suggesting replacement with `BackpressureDecompressionHandler`
* Create a replacement for `HttpContentDecoder` based on the new decompressors

### CODE REVIEW

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    private boolean downstreamWantsMore;
+    private boolean reading;
+
+    /**
+     * Create a new handler using the given decompressor.
+     *
+     * @param decompressor The decompressor
+     */
+    public BackpressureDecompressionHandler(Decompressor decompressor) {
+        this.decompressor = decompressor;
```
Comment (normanmaurer) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: missing null check

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        try {
+            return decompressor.status();
+        } catch (Exception e) {
+            handleException(e);
+            throw e;
+        }
+    }
+
+    private boolean downstreamWantsMore(ChannelHandlerContext ctx) {
+        return downstreamWantsMore || ctx.channel().config().isAutoRead();
```
Comment (normanmaurer) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: I wonder if we should respect `MESSAGES_PER_READ` here .. like read X number of times

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                        throw e;
+                    }
+                    if (!buf.isReadable()) {
+                        // try again.
+                        buf.release();
+                        break;
+                    }
+                    downstreamWantsMore = false;
+                    ctx.fireChannelRead(buf);
+                    ctx.fireChannelReadComplete();
```
Comment (normanmaurer) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: This is really wasteful imho... We should only call `fireChannelReadComplete()` once per `read()` as people use `channelReadComplete()` to flush and so do a `writev` for example

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    @Override
+    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+        reading = false;
+        if (!ctx.channel().config().isAutoRead() && downstreamWantsMore) {
+            ctx.read();
+        }
+    }
+
```
Comment (normanmaurer) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java:76]: I think we also need to add an override for `handlerAdded` and handle the `inputBuffer` there.

File: codec-compression/src/main/java/io/netty/handler/codec/compression/Decompressor.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+         * are permitted.
+         */
+        COMPLETE,
+    }
+
+    abstract class AbstractDecompressorBuilder {
+        protected final ByteBufAllocator allocator;
+
+        protected AbstractDecompressorBuilder(ByteBufAllocator allocator) {
+            this.allocator = allocator;
```
Comment (chrisvest) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/Decompressor.java]: Require non-null

File: codec-compression/src/main/java/io/netty/handler/codec/compression/DefensiveDecompressor.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import io.netty.buffer.ByteBuf;
+
+/**
+ * Most decompressor implementations play fast and loose with {@link Decompressor} API contracts. This wrapper makes
```
Comment (chrisvest) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/DefensiveDecompressor.java:22]: Are we questioning the Decompressor callers, or the Decompressor implementations? If the latter, can we cover that with tests instead?

File: codec-compression/src/test/java/io/netty/handler/codec/compression/Bzip2DecoderTest.java [hunk 7]
```diff
@@ -50,12 +51,7 @@ private void writeInboundDestroyAndExpectDecompressionException(ByteBuf in) {
         try {
             channel.writeInbound(in);
         } finally {
-            try {
-                destroyChannel();
-                fail();
```
Comment (chrisvest) [inline codec-compression/src/test/java/io/netty/handler/codec/compression/Bzip2DecoderTest.java:59]: So previously, `destroyChannel` was expected to also throw here. I can't but wonder what the idea was. And in this PR we have no changes to either `AbstractDecoderTest` or `Bzip2Decoder`?

No wait, `destroyChannel` still throws, but now that exception overwrites whatever was or wasn't thrown by `channel.writeInbound(in)`. So we are changing what we are asserting here. Previously we were asserting that `writeInbound` throws, but now we don't check that.

File: codec-compression/src/main/java/io/netty/handler/codec/compression/Decompressor.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+     * throw an exception if the input is truncated, but most implementations do nothing. Only permitted for
+     * {@link Status#NEED_INPUT}.
+     */
+    void endOfInput() throws DecompressionException;
+
+    /**
+     * Take a decompressed buffer from this decompressor. Only permitted for {@link Status#NEED_OUTPUT}. Buffer
+     * ownership transfers to the caller.
+     *
+     * @return The decompressed buffer. May be empty
```
Comment (chrisvest) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/Decompressor.java]: Missing `.` after "empty"

File: codec-compression/src/main/java/io/netty/handler/codec/compression/DefensiveDecompressor.java (hunk 6 above)
Comment (yawkat) [reply codec-compression/src/main/java/io/netty/handler/codec/compression/DefensiveDecompressor.java:22]: The Decompressor implementations do not properly check how Decompressor is called, e.g. that addInput is only called when status == NEED_INPUT. For some decompressors this would be fine, but for others it is not.

The only callers right now do follow the contract, but DefensiveDecompressor adds additional checks to make sure it stays that way, especially since these APIs are open to users.

File: codec-compression/src/test/java/io/netty/handler/codec/compression/Bzip2DecoderTest.java (hunk 7 above)
Comment (yawkat) [reply codec-compression/src/test/java/io/netty/handler/codec/compression/Bzip2DecoderTest.java:59]: iirc it throws before destroyChannel now, in the try block that's still there. The old code asserted that specifically the destroy threw, the new code doesn't care if it's the read or the destroy.

This change is necessary because I extend Bzip2DecoderTest to also run with the backpressure decoder.

File: codec-compression/src/main/java/io/netty/handler/codec/compression/Decompressor.java (hunk 8 above)
Comment (yawkat) [reply codec-compression/src/main/java/io/netty/handler/codec/compression/Decompressor.java]:
```suggestion
     * @return The decompressed buffer. May be empty.
```

**Review by chrisvest — APPROVED** [review]

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.internal.RecyclableArrayList;
+
+/**
+ * A channel handler that decompresses incoming {@link ByteBuf}s using a {@link Decompressor}. This is designed as a
+ * drop-in replacement for the various decompression decoders (e.g. {@link BrotliDecoder}), but properly handles
+ * downstream backpressure.
+ */
+public final class BackpressureDecompressionHandler extends ChannelDuplexHandler {
```
Comment (normanmaurer) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: Should we deprecate the old decoders and point to this one ?

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        }
+        processSome(ctx);
+    }
+
+    @Override
+    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+        reading = false;
+        if (!ctx.channel().config().isAutoRead() && downstreamWantsMore) {
+            ctx.read();
+        }
```
Comment (normanmaurer) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: Like mentioned above we need to do something more smart about `channelReadComplete` and so might need to also handle it here.

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BrotliDecompressor.java [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        }
+
+        /**
+         * Desired size of the input buffer in bytes. Default 8K.
+         *
+         * @param inputBufferSize desired size of the input buffer in bytes
+         * @return This builder
+         */
+        public Builder inputBufferSize(int inputBufferSize) {
+            this.inputBufferSize = inputBufferSize;
```
Comment (normanmaurer) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/BrotliDecompressor.java]: do we need to validate that this is non-negative ?

File: codec-compression/src/main/java/io/netty/handler/codec/compression/Bzip2Decompressor.java [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+        }
+
+        /**
+         * Size of the output buffer to return from {@link #takeOutput()}. Default 64K.
+         *
+         * @param outputBufferSize Output buffer size
+         * @return This builder
+         */
+        public Builder outputBufferSize(int outputBufferSize) {
+            this.outputBufferSize = outputBufferSize;
```
Comment (normanmaurer) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/Bzip2Decompressor.java]: same comment as above... validate ?

File: codec-compression/src/main/java/io/netty/handler/codec/compression/DefensiveDecompressor.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+ * sure callers follow that contract.
+ */
+final class DefensiveDecompressor implements Decompressor {
+    private final Decompressor delegate;
+    private Status status;
+    private boolean closed;
+    private boolean failed;
+
+    DefensiveDecompressor(Decompressor delegate) {
+        this.delegate = delegate;
```
Comment (normanmaurer) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/DefensiveDecompressor.java]: null check ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java (hunk 9 above)
Comment (yawkat) [reply codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: Probably yes… but I think it's too early to do this while HttpContentDecompressor does not have a replacement and is not deprecated

Comment (yawkat) [reply codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: Thinking about it, a HttpContentDecompressor replacement would probably share a lot of code with this class, so maybe I should refactor this a bit with to have a shared backpressure helper class that both can use.

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java (hunk 2 above)
Comment (yawkat) [reply codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: MAX_MESSAGES_PER_READ is deprecated and the replacement seems unsuited. Maybe there could be a property on this handler class though

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java (hunk 3 above)
Comment (yawkat) [reply codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: I think the behavior here is similar to FlowControlHandler, the best we could do to resolve this is your messages-per-read idea

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java (hunk 4 above)
Comment (yawkat) [reply codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java:76]: inputBuffer is only used when a decompress call makes produces more than one output buffer, which should not be the norm.

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java (hunk 3 above)
Comment (normanmaurer) [reply codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: Yeah maybe... I am just saying this will hurt performance a lot and so we need to do something about it. Beside the problem with flushing it also means we now need to traverse the pipeline twice per message which is bad even without the flushing problem.

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java (hunk 2 above)
Comment (normanmaurer) [reply codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: We should most likely just `de-deprecate` it btw.

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BrotliDecompressor.java (hunk 11 above)
Comment (yawkat) [reply codec-compression/src/main/java/io/netty/handler/codec/compression/BrotliDecompressor.java]: i'll add a check, I assume otherwise the native code checks this

File: codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java (hunk 3 above)
Comment (yawkat) [reply codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: Well in the ideal case there's exactly one decompressed buffer per incoming compressed buffer. in that case the pipeline traversals should be unchanged (we eat the upstream channelReadComplete, but produce our own). I'm not sure if one-to-many decompression is common in practice

Comment (normanmaurer) [reply codec-compression/src/main/java/io/netty/handler/codec/compression/BackpressureDecompressionHandler.java]: while its true there will be most likely a buffer to buffer mapping its still true that you will most likely receive multiple buffers from the transport before you finally receive the fireChannelReadComplete() so my concern still is valid. Like usually what happens with for example epoll is that we will read like 16 buffers and so fire 16 buffers through the pipeline before we fire one channelReadComplete.

**Comment by yawkat** [comment]
@normanmaurer I've added new logic that allows a number of buffers to be forwarded before a readComplete, and a test. I hope that works for you. The logic is getting complicated though. PTAL.

I've also changed the handler API to create the decompressor lazily, in handlerAdded. This looks a bit odd (you pass the decompressor builder, not the decompressor, to the handler) but I think it works as an API and it's safer.

File: codec-http/src/main/java/io/netty/handler/codec/http/BackpressureHttpContentDecoder.java [hunk 14]
```diff
@@ -0,0 +1,564 @@
...
+                    decoder.channelInactive(decoderContext);
+                    cleanup();
+
+                    LastHttpContent last = (LastHttpContent) c;
+                    // Generate an additional chunk if the decoder produced
+                    // the last product on closure,
+                    HttpHeaders headers = last.trailingHeaders();
+                    needRead = false;
+                    if (headers.isEmpty()) {
+                        ctx.fireChannelRead(LastHttpContent.EMPTY_LAST_CONTENT);
```
Comment (yawkat) [inline codec-http/src/main/java/io/netty/handler/codec/http/BackpressureHttpContentDecoder.java]: just realized this is wrong, it needs to be delayed further. i will fix it tomorrow.

**Comment by m-burst** [comment]
Hey folks, do you think this will get merged anytime soon?

File: codec-compression/src/main/java/io/netty/handler/codec/compression/Lz4FrameDecompressor.java [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+            throw t;
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final class Builder extends AbstractDecompressorBuilder {
+        private LZ4Factory factory = LZ4Factory.fastestJavaInstance();
```
Comment (yawkat) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/Lz4FrameDecompressor.java]:
```suggestion
        private LZ4Factory factory = LZ4Factory.fastestInstance();
```

File: codec-compression/src/main/java/io/netty/handler/codec/compression/Lz4FrameDecompressor.java [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+        DECOMPRESS_DATA,
+        FINISHED,
+    }
+
+    private State currentState = State.INIT_BLOCK;
+
+    /**
+     * Underlying decompressor in use.
+     */
+    private LZ4FastDecompressor decompressor;
```
Comment (yawkat) [inline codec-compression/src/main/java/io/netty/handler/codec/compression/Lz4FrameDecompressor.java]: need to move to LZ4SafeDecompressor here

**Comment by chrisvest** [comment]
There's also merge conflicts that need to be fixed. Please ping me when I should take a look again.

**Comment by yawkat** [comment]
@chrisvest updated

**Comment by yawkat** [comment]
I found one likely regression in the new HTTP/2 decompression path:

In `codec-http2/src/main/java/io/netty/handler/codec/http2/DecompressingFrameListener.java`, `consumeBytes(...)` ultimately has to return the compressed bytes charged against HTTP/2 flow control, which includes frame padding as well as `data.readableBytes()`. The old `DelegatingDecompressorFrameListener` implementation included padding in its compressed/decompressed accounting, but here `StreamHolder` only tracks `buffer.readableBytes()` and always forwards decompressed output with `padding = 0`.

For padded DATA frames, those padding bytes never seem to get returned to the upstream flow controller. Under enough padded traffic, that would permanently shrink the local receive window and can stall the stream/connection.

I think this needs to preserve padding in the accounting path, and it would be good to add a test that exercises non-zero padding, since the new `DecompressingFrameListenerTest` currently only covers `padding = 0`.
